sklearn/inspection/_partial_dependence.py

```diff
@@ -79,30 +77,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
 
     values = []
     for feature in range(X.shape[1]):
-        uniques = np.unique(_safe_indexing(X, feature, axis=1))
-        if uniques.shape[0] < grid_resolution:
-            # feature has low resolution use unique vals
-            axis = uniques
-        else:
-            # create axis based on percentiles and grid resolution
-            emp_percentiles = mquantiles(
-                _safe_indexing(X, feature, axis=1), prob=percentiles, axis=0
-            )
-            if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+        if feature in custom_values:
+            # Use values in the custom range
+            feature_range = custom_values[feature]
+            if not isinstance(feature_range, np.ndarray):
+                feature_range = np.array(feature_range)
+            if feature_range.ndim != 1:
                 raise ValueError(
-                    "percentiles are too close to each other, "
-                    "unable to build the grid. Please choose percentiles "
-                    "that are further apart."
+                    "Grid for feature {} is not a one-dimensional array. Got {}"
+                    " dimensions".format(feature, feature_range.ndim)
+                )
+            axis = feature_range
+        else:
+            uniques = np.unique(_safe_indexing(X, feature, axis=1))
+            if uniques.shape[0] < grid_resolution:
+                # feature has low resolution use unique vals
+                axis = uniques
+            else:
+                # create axis based on percentiles and grid resolution
+                emp_percentiles = mquantiles(
+                    _safe_indexing(X, feature, axis=1), prob=percentiles, axis=0
+                )
+                if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+                    raise ValueError(
+                        "percentiles are too close to each other, "
+                        "unable to build the grid. Please choose percentiles "
+                        "that are further apart."
+                    )
+                axis = np.linspace(
+                    emp_percentiles[0],
+                    emp_percentiles[1],
+                    num=grid_resolution,
+                    endpoint=True,
                 )
-            axis = np.linspace(
-                emp_percentiles[0],
-                emp_percentiles[1],
-                num=grid_resolution,
-                endpoint=True,
-            )
         values.append(axis)
 
-    return cartesian(values), values
+    # Create a place holder for the cartesian product of the individual grids.
+    shape = (len(v) for v in values)
+    ix = np.indices(shape)
```

Comment:
It's probably a bad idea to create an array with a new dimension for every single feature, because this line will fail for >32 features.

```
Traceback (most recent call last):
  File ".../lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py", line 521, in partial_dependence
    custom_values_idx
  File ".../lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py", line 120, in _grid_from_X
    ix = np.indices(shape)
  File ".../lib/python3.7/site-packages/numpy/core/numeric.py", line 1777, in indices
    res = empty((N,)+dimensions, dtype=dtype)
ValueError: maximum supported dimension for an ndarray is 32, found 792
```

Reply:
You're correct but that limitation exists on the current main branch

```python
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
from sklearn.inspection import partial_dependence
import pytest

data = load_digits()
X = data.data
y = data.target

rf = RandomForestClassifier()
rf.fit(X, y)

with pytest.raises(ValueError, match="maximum supported dimension for an ndarray is 32, found 33"):
    partial_dependence(rf, X, features=list(range(32)))
```

So if we want to support more than 32-way partial dependence, that's a separate ticket.